db/dohast.c

```diff
@@ -0,0 +1,657 @@
+{
+    int ignore = 0;
+
+    if (bdb_attr_get(thedb->bdb_attr, BDB_ATTR_DOHAST_DISABLE))
+        return 0;
+
+    if (dohsql_is_parallel_shard())
+        return 0;
+
+    if (ast->nused >= ast->nalloc) {
+        ast->stack = realloc(ast->stack, (ast->nalloc + AST_STACK_INIT) *
+                                             sizeof(ast->stack[0]));
+        if (!ast->stack)
+            return -1;
+        bzero(&ast->stack[ast->nalloc], AST_STACK_INIT * sizeof(ast->stack[0]));
+        ast->nalloc += AST_STACK_INIT;
+    }
+
+    switch (op) {
+    case AST_TYPE_SELECT: {
+        Select *p = (Select *)obj;
+
+        if ((p->selFlags & SF_ASTIncluded) == 0) {
+            ast->stack[ast->nused].op = op;
+            ast->stack[ast->nused].obj = gen_select(v, p);
+            ast->nused++;
+        } else {
+            ignore = 1;
+        }
+        break;
+    }
+    default: {
+        ast->stack[ast->nused].op = op;
+        ast->stack[ast->nused].obj = obj;
+        ast->nused++;
+        break;
+    }
+    }
+
+    if (ast_verbose && !ignore) {
+        ast_print(ast);
+    }
+
+    return 0;
+}
+
+int ast_pop(ast_t *ast, ast_node_t **pnode)
+{
+    if (ast->nused > 0) {
+        ast->nused--;
+        *pnode = &ast->stack[ast->nused];
+    } else
+        return -1;
+
+    return ast->nused;
+}
+
+const char *ast_type_str(enum ast_type type)
+{
+    switch (type) {
+    case AST_TYPE_SELECT:
+        return "SELECT";
+    case AST_TYPE_INSERT:
+        return "INSERT";
+    case AST_TYPE_IN:
+        return "IN_SELECT";
+    case AST_TYPE_DELETE:
+        return "DELETE";
+    case AST_TYPE_UPDATE:
+        return "UPDATE";
+    default:
+        return "INVALID";
+    }
+}
+
+const char *ast_param_str(enum ast_type type, void *obj)
+{
+    switch (type) {
+    case AST_TYPE_SELECT:
+        return obj ? ((dohsql_node_t *)obj)->sql : "NULL";
+    case AST_TYPE_INSERT:
+        return "()";
+    case AST_TYPE_IN:
+        return "()";
+    case AST_TYPE_DELETE:
+        return "()";
+    case AST_TYPE_UPDATE:
+        return "()";
+    default:
+        return "INVALID";
+    }
+}
+
+void ast_print(ast_t *ast)
+{
+    int i;
+    logmsg(LOGMSG_ERROR, "AST: [%d]\n", ast->nused);
+    for (i = 0; i < ast->nused; i++)
+        logmsg(LOGMSG_ERROR, "\t %d. %s \"%s\"\n", i,
+               ast_type_str(ast->stack[i].op),
+               ast_param_str(ast->stack[i].op, ast->stack[i].obj));
+}
+
+int comdb2_check_parallel(Parse *pParse)
+{
+    ast_t *ast = pParse->ast;
+    dohsql_node_t *node;
+    int i;
+
+    if (bdb_attr_get(thedb->bdb_attr, BDB_ATTR_DOHSQL_DISABLE))
+        return 0;
+
+    if (has_parallel_sql(NULL) == 0)
+        return 0;
+
+    if (ast->nused > 1)
+        return 0;
+
+    if (ast->stack[0].op != AST_TYPE_SELECT)
+        return 0;
+    if (!ast->stack[0].obj)
+        return 0;
+
+    node = (dohsql_node_t *)ast->stack[0].obj;
+
+    if (strstr(node->sql, "sqlite_"))
```

Comment:
strcasestr (or really the parser should mark it as ineligible).

Reply:
This will be reviewed on a further iteration; for now, to avoid internally generated queries that nest with the user queries, I am simply skipping the parallel optimization for these statements .

Reply:
switched to case insensitive